File: backend/app/services/zapsign.py
import os
import base64
import requests
import logging

ZAPSIGN_API_KEY = os.getenv("ZAPSIGN_API_KEY", "")
ZAPSIGN_API_URL = "https://api.zapsign.com.br/api/v1/docs"

logger = logging.getLogger("uvicorn")

def create_zapsign_document(
    nome_completo: str,
    email: str,
    nome_comercial: str,
    contrato_path: str
) -> dict:
    """
    Uploads the compiled contract to ZapSign and creates a document for signing.
    If ZAPSIGN_API_KEY is empty, it returns simulated sandbox URLs.
    """
    if not ZAPSIGN_API_KEY:
        logger.info("[ZapSign Integration] Running in Simulation Mode (No API key provided).")
        simulated_doc_id = "zap-sim-doc-123456789"
        simulated_sign_url = f"https://sandbox.zapsign.com.br/sign/{simulated_doc_id}"
        return {
            "status": "simulated",
            "doc_id": simulated_doc_id,
            "sign_url": simulated_sign_url
        }

    if not contrato_path or not os.path.exists(contrato_path):
        raise FileNotFoundError(f"Arquivo de contrato não encontrado para envio: {contrato_path}")

    # Read DOCX file and convert to Base64
    with open(contrato_path, "rb") as file:
        encoded_string = base64.b64encode(file.read()).decode("utf-8")

    filename = os.path.basename(contrato_path)

    # Prepare ZapSign Payload
    # API Docs: https://docs.zapsign.com.br/
    payload = {
        "name": f"Contrato Cosmo Alma TV - {nome_comercial}",
        "base64_pdf": f"data:application/vnd.openxmlformats-officedocument.wordprocessingml.document;base64,{encoded_string}",
        "signers": [
            {
                "name": nome_completo,
                "email": email,
                "send_email": True,
                "auth_mode": "signature"
            }
        ]
    }

    headers = {
        "Authorization": f"Bearer {ZAPSIGN_API_KEY}",
        "Content-Type": "application/json"
    }

    try:
        response = requests.post(ZAPSIGN_API_URL, json=payload, headers=headers)
        response.raise_for_status()
        data = response.json()

        # Extract document and signer details
        doc_id = data.get("token")
        signers = data.get("signers", [])
        sign_url = signers[0].get("sign_url") if signers else ""

        logger.info(f"[ZapSign Integration] Document created successfully. ID: {doc_id}")
        return {
            "status": "success",
            "doc_id": doc_id,
            "sign_url": sign_url
        }
    except Exception as e:
        logger.error(f"[ZapSign Integration] Error creating document on ZapSign: {str(e)}")
        raise e

File: src/lib/services/zapsign.ts
const ZAPSIGN_API_KEY = process.env.ZAPSIGN_API_KEY || "";
const ZAPSIGN_API_URL = "https://api.zapsign.com.br/api/v1/docs";

export async function createZapSignDocument(
  nomeCompleto: string,
  email: string,
  nomeComercial: string,
  contratoBase64: string, // We pass base64 content directly instead of path
  filename: string
): Promise<{ status: string; doc_id: string; sign_url: string }> {
  if (!ZAPSIGN_API_KEY) {
    console.warn("[ZapSign Integration] Running in Simulation Mode (No API key provided).");
    const simulatedDocId = "zap-sim-doc-123456789";
    const simulatedSignUrl = `https://sandbox.zapsign.com.br/sign/${simulatedDocId}`;
    return {
      status: "simulated",
      doc_id: simulatedDocId,
      sign_url: simulatedSignUrl,
    };
  }

  const payload = {
    name: `Contrato Cosmo Alma TV - ${nomeComercial}`,
    base64_pdf: `data:application/vnd.openxmlformats-officedocument.wordprocessingml.document;base64,${contratoBase64}`,
    signers: [
      {
        name: nomeCompleto,
        email: email,
        send_email: true,
        auth_mode: "signature",
      },
    ],
  };

  const response = await fetch(ZAPSIGN_API_URL, {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${ZAPSIGN_API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorText = await response.text();
    console.error(`[ZapSign Integration] Error creating document: ${errorText}`);
    throw new Error(`ZapSign API Error: ${errorText}`);
  }

  const data = await response.json();
  const docId = data.token;
  const signers = data.signers || [];
  const signUrl = signers[0]?.sign_url || "";

  return {
    status: "success",
    doc_id: docId,
    sign_url: signUrl,
  };
}
